scheduler/job.c

```diff
@@ -3736,7 +3746,72 @@ get_options(cupsd_job_t *job,		/* I - Job */
       {
         if (!ippFindAttribute(job->attrs, preset->name, IPP_TAG_ZERO))
         {
-          cupsdLogJob(job, CUPSD_LOG_DEBUG2, "Adding preset option %s=%s", preset->name, preset->value);
+          cupsdLogJob(job, CUPSD_LOG_DEBUG, "Adding preset option %s=%s", preset->name, preset->value);
+
+	  num_pwgppds = cupsAddOption(preset->name, preset->value, num_pwgppds, &pwgppds);
+        }
+      }
+    }
+  }
+
+  if (pc &&
+      ippFindAttribute(job->attrs, "print-content-optimize", IPP_TAG_ZERO))
+  {
+   /*
+    * Map print-content-optimize to a preset...
+    */
+
+    if ((attr = ippFindAttribute(job->attrs, "print-content-optimize",
+				 IPP_TAG_KEYWORD)) != NULL)
+    {
+      if (!strcmp(attr->values[0].string.text, "auto"))
+	print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_AUTO;
+      else if (!strcmp(attr->values[0].string.text, "photo"))
+	print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_PHOTO;
+      else if (!strcmp(attr->values[0].string.text, "graphics") ||
+	       !strcmp(attr->values[0].string.text, "graphic"))
+	print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_GRAPHICS;
+      else if (!strcmp(attr->values[0].string.text, "text"))
+	print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_TEXT;
+      else if (!strcmp(attr->values[0].string.text, "text-and-graphics") ||
+	       !strcmp(attr->values[0].string.text, "text-and-graphic"))
+	print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_TEXT_AND_GRAPHICS;
+      else
+	print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_AUTO;
+    }
+    else
+      print_content_optimize = _PWG_PRINT_CONTENT_OPTIMIZE_AUTO;
+
+  cupsdLogJob(job, CUPSD_LOG_DEBUG,
+	      "print-content-optimize=%s",
+              (print_content_optimize == _PWG_PRINT_CONTENT_OPTIMIZE_AUTO ?
+	       "automatic" :
+               (print_content_optimize == _PWG_PRINT_CONTENT_OPTIMIZE_PHOTO ?
+		"photo" :
+                (print_content_optimize == _PWG_PRINT_CONTENT_OPTIMIZE_GRAPHICS ?
+		 "graphics" :
+                 (print_content_optimize == _PWG_PRINT_CONTENT_OPTIMIZE_TEXT ?
+		  "text" :
+                  "text and graphics")))));
+    if (pc->num_optimize_presets[print_content_optimize] > 0)
+    {
+     /*
+      * Copy the preset options as long as the corresponding names are not
+      * already defined in the IPP request and also if it does not change
+      * the print quality preset (as long as we do not print in high quality)
+      * ...
+      */
+
+      for (i = pc->num_optimize_presets[print_content_optimize],
+	       preset = pc->optimize_presets[print_content_optimize];
+	   i > 0;
+	   i --, preset ++)
+      {
+        if (!ippFindAttribute(job->attrs, preset->name, IPP_TAG_ZERO) &&
```

Comment:
What are you trying to do here? Better to:

a. Define a cupsPrintOptimize option in the PPD
b. Include "*cupsPrintOptimize Photo" for the photo preset

Then we don't have any special casing.

Reply:
As I told in the answer of your comment above, I do not want to modify PPD files but have a solution which works at run-time and fully automatically for the user, on existing PPDs of non-driverless printers.

Reply:
I am not asking you to modify PPD files.

*If* you are mapping vendor PPD options to the "print-content-optimize" attribute, then you can reflect this in the PPD cache like we do for "sides". The auto-generated IPP Everywhere PPD can use the name "cupsPrintOptimize" for the PPD option which would be one of the supported PPD option keywords that get mapped to "print-content-optimize".

What I'm thinking is that you'd add the following to the `_ppd_cache_s` structure in ppd-private.h:

```
char *optimize_option, /* PPD option for print-content-optimize */
     *optimize_values[_PWG_OPTIMIZE_MAX]; /* PPD choices for print-content-optimize values */
```

Reply:
My approach allows mapping `print-content-optimize` to more than one single option in the PPD file, so it is more universally applicable.
We should give this array of five option/value lists a better name than `optimizePresets`.

Reply:
Till, unless there are specific PPDs you plan to support, we don't need the complexity. And honestly we barely need print-content-optimize - the only useful value is 'photo' which can normally be inferred from the content. The the class of printer you are supporting via PPD files do *not* offer photo quality... :/

Reply:
I simply saw this IPP attribute put prominently into the web interface of PAPPL ("Prtinting Defaults") and so I thought that there are many (non-driverless, these are the ones PAPPL is for) printers somehow supporting it. And I also saw in PPD files of non-driverless printers options for content optimization, with not necessarily having the same names at every driver/manufacturer, so I tried to map these to the `print-content-optimize` attribute to make it useful, especially as I created this code originally for retro-fitting PPD-based drivers (and PostScript PPDs) into PAPPL-based Printer Applications. Then later it came up the idea in me to use that code in CUPS.

Printers with PPD files are not only PostScript laser printers, there are also inkjets supported with HPLIP, Gutenprint, ... These do photo quality and they are controlled by PPD files. So users currently set them up with a PPD-file-based driver in CUPS.

I know that Gutenprint and HPLIP are actively maintained and so their maintainers should create native Printer Applications out of them, but this can take time, so I have covered them in my retro-fitting Printer Applications to work the best possible for everyone. I think to be able to convince the HPLIP guys to switch over scanning support in PAPPL needs to get finished.

Reply:
And by the way, in the PAPPL retro-fit library I try to auto-detect print job content, especially image file input (JPG or PNG) is considered photo and if no specific print-content optimization is selected I select "photo" then.